File: problems/Python/87. Scramble String.py
class Solution:
    def isScramble(self, s1: str, s2: str) -> bool:
        memo = {}
        def solve(a, b):
            if (a, b) in memo:
                return memo[(a, b)]
            if a == b:
                return True
            if sorted(a) != sorted(b):
                return False

            n = len(a)
            for i in range(1, n):
                no_swap = solve(a[:i], b[:i]) and solve(a[i:], b[i:])
                if no_swap:
                    memo[(a, b)] = True
                    return True
                
                swap = solve(a[:i], b[n-i:]) and solve(a[i:], b[:n-i])
                if swap:
                    memo[(a, b)] = True
                    return True
            
            memo[(a, b)] = False
            return False

        return solve(s1, s2)

File: problems/C++/87. Scramble String.cpp
class Solution {
public:
    unordered_map<string, bool> memo;

    bool isScramble(string s1, string s2) {
        string key = s1 + "#" + s2;
        if (memo.count(key)) return memo[key];
        if (s1 == s2) return true;
        
        string temp1 = s1, temp2 = s2;
        sort(temp1.begin(), temp1.end());
        sort(temp2.begin(), temp2.end());
        if (temp1 != temp2) return false;

        int n = s1.length();
        for (int i = 1; i < n; ++i) {
            bool no_swap = isScramble(s1.substr(0, i), s2.substr(0, i)) &&
                           isScramble(s1.substr(i), s2.substr(i));
            if (no_swap) return memo[key] = true;
            
            bool swap = isScramble(s1.substr(0, i), s2.substr(n - i)) &&
                        isScramble(s1.substr(i), s2.substr(0, n - i));
            if (swap) return memo[key] = true;
        }

        return memo[key] = false;
    }
};